TrustedTypes: Worker()/SharedWorker() constructors (#42073)
mdn/content: files/en-us/web/api/sharedworker/sharedworker/index.md

{{APIRef("Web Workers API")}}

> [!WARNING]
> This script passed to the `url` element is executed.
> APIs like this are known as [injection sinks](/en-US/docs/Web/API/Trusted_Types_API#concepts_and_usage), and are potentially a vector for [cross-site scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.
>
> You can mitigate this risk by having a [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) that restricts the locations from which scripts can be loaded, and by always assigning {{domxref("TrustedScriptURL")}} objects instead of strings and [enforcing trusted types](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types).
> See [Security considerations](/en-US/docs/Web/API/Worker/Worker#security_considerations) in the `Worker()` constructor for more information.

The **`SharedWorker()`** constructor creates a {{domxref("SharedWorker")}} object that executes the script at the specified URL.

> This script must obey the [same-origin policy](/en-US/docs/Web/Security/Defenses/Same-origin_policy).

Author: Yes. I'll update as you suggest. - I'm so exhausted by the testing.

### Parameters

- `url`
  - : A {{domxref("TrustedScriptURL")}} object or a string representing the URL of the script or module that the worker will execute.

>     This must be same-origin with the caller's document, a `blob:`, or a `data:` URL.

Author: ```suggestion
    This must be same-origin with the caller's document, or a `blob:` or `data:` URL.
```

>     This must be same-origin with the associated document (`blob:` URLs are same-origin), or a `data:` URL.

Author: Yes it is

>     This must be same-origin with the associated document (`blob:` URLs are same-origin), or a `data:` URL.

Author: ```suggestion
    This must be same-origin with the caller's document, a `blob:`, or a `data:` URL.
```

        Can have one of the following two values:
        - 'all'
          - : `SameSite=Strict`, `SameSite=Lax`, and `SameSite=None` cookies will all be available to the worker.
            This option is only supported in first-party contexts, and is the default in first-party contexts.
        - 'none'
          - : Only `SameSite=None` cookies will be available to the worker.
            This option is supported in first-party and third-party contexts, and is the default in third-party contexts.

### Exceptions

- `SecurityError` {{domxref("DOMException")}}
  - : Thrown if the document is not allowed to start workers, for example if the URL has an invalid syntax or if the same-origin policy is violated, or if the `sameSiteCookies` value is not supported in the given context.
- `NetworkError` {{domxref("DOMException")}}
  - : Thrown if the MIME type of the worker script is incorrect.
    It should _always_ be `text/javascript` (for historical reasons [other JavaScript MIME types](/en-US/docs/Web/HTTP/Guides/MIME_types#textjavascript) may be accepted).
- `SyntaxError` {{domxref("DOMException")}}
  - : Thrown if `url` cannot be parsed.
- `TypeError`

>   - : Thrown if the `url` parameter is set with a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.

Reviewer: ```suggestion
  - : Thrown if the `url` parameter is a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.
```
